File: packages/soulseek_protocol/lib/src/network/distributed_network.dart
import 'dart:async';

import '../connection/socket_manager.dart';
import '../messages/buffer.dart';
import '../messages/codes.dart';
import '../services/search_service.dart';

enum DistributedState { disconnected, connecting, connected, disconnecting }

enum DistributedRole { none, parent, child }

class DistributedNetwork {
  SocketTransport? _parentSocket;
  DistributedState _state = DistributedState.disconnected;
  DistributedRole _role = DistributedRole.none;
  final _searchRequestsController = StreamController<DistributedSearchRequest>.broadcast();
  final _stateController = StreamController<DistributedState>.broadcast();
  StreamSubscription? _parentMessageSub;

  final String username;
  final int _maxChildBranches = 10;
  final List<SocketTransport> _childConnections = [];
  final SocketTransport Function() _socketFactory;

  DistributedNetwork({required this.username, SocketTransport Function()? socketFactory})
      : _socketFactory = socketFactory ?? (() => SocketManager());

  Stream<DistributedState> get stateChanges => _stateController.stream;
  Stream<DistributedSearchRequest> get searchRequests => _searchRequestsController.stream;
  DistributedState get state => _state;
  DistributedRole get role => _role;
  int get childCount => _childConnections.length;

  Future<bool> connectToParent(String host, int port) async {
    _setState(DistributedState.connecting);
    final socket = _socketFactory();
    try {
      await socket.connect(host, port);
      if (socket.state != SocketState.connected) {
        socket.dispose();
        _setState(DistributedState.disconnected);
        return false;
      }
      _parentSocket = socket;
      _role = DistributedRole.child;
      _parentMessageSub = socket.messages.listen(_onParentMessage);
      _setState(DistributedState.connected);
      return true;
    } catch (_) {
      socket.dispose();
      _setState(DistributedState.disconnected);
      return false;
    }
  }

  void becomeParent() {
    _role = DistributedRole.parent;
    _setState(DistributedState.connected);
  }

  bool acceptChildConnection(SocketTransport socket) {
    if (_childConnections.length >= _maxChildBranches) return false;
    _childConnections.add(socket);
    if (_role == DistributedRole.none) {
      _role = DistributedRole.parent;
    }
    return true;
  }

  void relaySearchRequest(String query, int ticket, {SocketTransport? exclude}) {
    if (_role != DistributedRole.parent) return;

    final w = WriteBuffer();
    w.writeString(username);
    w.writeInt32(ticket);
    w.writeString(query);

    for (final child in _childConnections) {
      if (child == exclude) continue;
      child.sendMessage(SoulseekMessage(DistantCode.fileSearch, w.toBytes()));
    }
  }

  void relaySearchResponse(SearchResult result, {SocketTransport? exclude}) {
    if (_role != DistributedRole.child || _parentSocket == null) return;

    final w = WriteBuffer();
    w.writeString(result.username);
    w.writeInt32(result.ticket);
    w.writeInt32(result.freeUploadSlots);
    w.writeInt32(result.uploadSpeed);
    w.writeInt32(result.queueLength);
    w.writeInt32(result.files.length);

    for (final file in result.files) {
      w.writeInt32(file.code);
      w.writeString(file.filename);
      w.writeInt32(file.size);
      w.writeString(file.extension);
      w.writeInt32(file.attributeCount);
      w.writeInt32(0); w.writeInt32(file.bitrate);
      w.writeInt32(1); w.writeInt32(file.duration);
      w.writeInt32(2); w.writeInt32(file.sampleRate);
    }

    _parentSocket!.sendMessage(SoulseekMessage(DistantCode.fileSearchReply, w.toBytes()));
  }

  void _onParentMessage(SoulseekMessage message) {
    if (message.code != DistantCode.fileSearch) return;

    try {
      final buffer = ReadBuffer(message.payload);
      final remoteUsername = buffer.readString();
      final ticket = buffer.readInt32();
      final query = buffer.readString();

      _searchRequestsController.add(DistributedSearchRequest(
        username: remoteUsername,
        ticket: ticket,
        query: query,
      ));
    } catch (_) {
      // Skip malformed messages
    }
  }

  void _setState(DistributedState state) {
    if (_state == state) return;
    _state = state;
    if (!_stateController.isClosed) {
      _stateController.add(state);
    }
  }

  void disconnect() {
    _setState(DistributedState.disconnecting);
    _parentMessageSub?.cancel();
    _parentSocket?.disconnect();
    _parentSocket?.dispose();
    for (final child in _childConnections) {
      child.disconnect();
      child.dispose();
    }
    _childConnections.clear();
    _parentSocket = null;
    _role = DistributedRole.none;
    _setState(DistributedState.disconnected);
  }

  void dispose() {
    disconnect();
    if (!_searchRequestsController.isClosed) {
      _searchRequestsController.close();
    }
    if (!_stateController.isClosed) {
      _stateController.close();
    }
  }
}

class DistributedSearchRequest {
  final String username;
  final int ticket;
  final String query;

  const DistributedSearchRequest({
    required this.username,
    required this.ticket,
    required this.query,
  });
}

File: Soulseek.Protocol/Network/DistributedNetwork.cs
using Soulseek.Protocol.Connection;
using Soulseek.Protocol.Messages;
using Soulseek.Protocol.Services;

namespace Soulseek.Protocol.Network;

public enum DistributedState
{
    Disconnected,
    Connecting,
    Connected,
    Disconnecting
}

public enum DistributedRole
{
    None,
    Parent,
    Child
}

public record DistributedSearchRequest(string Username, int Ticket, string Query);

public class DistributedNetwork
{
    private ISocketTransport? _parentSocket;
    private DistributedState _state = DistributedState.Disconnected;
    private DistributedRole _role = DistributedRole.None;
    private readonly Subject<DistributedSearchRequest> _searchRequestsSubject = new();
    private readonly Subject<DistributedState> _stateSubject = new();
    private IDisposable? _parentMessageSub;

    public string Username { get; }
    public int MaxChildBranches { get; } = 10;
    private readonly List<ISocketTransport> _childConnections = [];
    private readonly Func<ISocketTransport> _socketFactory;

    public DistributedNetwork(string username, Func<ISocketTransport>? socketFactory = null)
    {
        Username = username;
        _socketFactory = socketFactory ?? (() => new SocketManager());
    }

    public IObservable<DistributedState> StateChanges => _stateSubject;
    public IObservable<DistributedSearchRequest> SearchRequests => _searchRequestsSubject;
    public DistributedState State => _state;
    public DistributedRole Role => _role;
    public int ChildCount => _childConnections.Count;

    public async Task<bool> ConnectToParent(string host, int port)
    {
        SetState(DistributedState.Connecting);
        var socket = _socketFactory();
        try
        {
            await socket.Connect(host, port);
            if (socket.State != SocketState.Connected)
            {
                socket.Dispose();
                SetState(DistributedState.Disconnected);
                return false;
            }
            _parentSocket = socket;
            _role = DistributedRole.Child;
            _parentMessageSub = socket.Messages.Subscribe(OnParentMessage);
            SetState(DistributedState.Connected);
            return true;
        }
        catch
        {
            socket.Dispose();
            SetState(DistributedState.Disconnected);
            return false;
        }
    }

    public void BecomeParent()
    {
        _role = DistributedRole.Parent;
        SetState(DistributedState.Connected);
    }

    public bool AcceptChildConnection(ISocketTransport socket)
    {
        if (_childConnections.Count >= MaxChildBranches) return false;
        _childConnections.Add(socket);
        if (_role == DistributedRole.None)
            _role = DistributedRole.Parent;
        return true;
    }

    public void RelaySearchRequest(string query, int ticket, ISocketTransport? exclude = null)
    {
        if (_role != DistributedRole.Parent) return;

        var w = new WriteBuffer();
        w.WriteString(Username);
        w.WriteInt32(ticket);
        w.WriteString(query);

        foreach (var child in _childConnections)
        {
            if (child == exclude) continue;
            child.SendMessage(new SoulseekMessage(DistantCode.FileSearch, w.ToBytes()));
        }
    }

    public void RelaySearchResponse(SearchResult result, ISocketTransport? exclude = null)
    {
        if (_role != DistributedRole.Child || _parentSocket == null) return;

        var w = new WriteBuffer();
        w.WriteString(result.Username);
        w.WriteInt32(result.Ticket);
        w.WriteInt32(result.FreeUploadSlots);
        w.WriteInt32(result.UploadSpeed);
        w.WriteInt32(result.QueueLength);
        w.WriteInt32(result.Files.Count);

        foreach (var file in result.Files)
        {
            w.WriteInt32(file.Code);
            w.WriteString(file.Filename);
            w.WriteInt32(file.Size);
            w.WriteString(file.Extension);
            w.WriteInt32(file.AttributeCount);
            w.WriteInt32(0); w.WriteInt32(file.Bitrate);
            w.WriteInt32(1); w.WriteInt32(file.Duration);
            w.WriteInt32(2); w.WriteInt32(file.SampleRate);
        }

        _parentSocket.SendMessage(new SoulseekMessage(DistantCode.FileSearchReply, w.ToBytes()));
    }

    private void OnParentMessage(SoulseekMessage message)
    {
        if (message.Code != DistantCode.FileSearch) return;

        try
        {
            var buffer = new ReadBuffer(message.Payload);
            var remoteUsername = buffer.ReadString();
            var ticket = buffer.ReadInt32();
            var query = buffer.ReadString();

            _searchRequestsSubject.OnNext(new DistributedSearchRequest(remoteUsername, ticket, query));
        }
        catch
        {
            // Skip malformed messages
        }
    }

    private void SetState(DistributedState state)
    {
        if (_state == state) return;
        _state = state;
        _stateSubject.OnNext(state);
    }

    public void Disconnect()
    {
        SetState(DistributedState.Disconnecting);
        _parentMessageSub?.Dispose();
        _parentSocket?.Disconnect();
        _parentSocket?.Dispose();
        foreach (var child in _childConnections)
        {
            child.Disconnect();
            child.Dispose();
        }
        _childConnections.Clear();
        _parentSocket = null;
        _role = DistributedRole.None;
        SetState(DistributedState.Disconnected);
    }

    public void Dispose()
    {
        Disconnect();
        _searchRequestsSubject.Dispose();
        _stateSubject.Dispose();
    }
}